ci/build.py

```diff
@@ -117,6 +123,9 @@ def run_cmd():
     image_id = _get_local_image_id(docker_tag=tag)
     if not image_id:
         raise FileNotFoundError('Unable to find docker image id matching with {}'.format(tag))
+    # now that we've built the container, push it to our docker cache if DOCKER_ECR_CACHE is defined
+    if 'DOCKER_ECR_REGISTRY' in os.environ:
+        push_docker_cache(registry, tag, image_id)
```

Comment:
This action breaks the security concept where only restricted jobs are allowed to publish caches for security reasons. Also, it results in a lot of unnecessary push queries to ECR during the PR jobs.

Reply:
Yes, it allows us to cache much more aggressively and eliminate the need for a job to build and push these images up. The ECR registry is not public readable, so it's not like we could potentially publish something bad. IMO, this is no different than using ccache for compiled files.

Reply:
Contributors can already modify the dockerfile's content in a PR which is immediately run, so having this false sense of security because only a special job can publish containers to a private registry is nonsense.

In addition, docker will not push layers that already exists in the registry, so the push essential becomes a no-op.

Reply:
The issue right now though is that with this design, you are granting regular untrusted slaves write-access to ECR. Unless you introduce another ECR repository for restricted jobs, this could be leveraged to modify the cache which is used by restricted jobs to inject malware into these jobs - e.g. when generating published artifacts.

Thus I'd recommend to either create two separate ECR repositories and only grant the appropriate Instance Profile the according permission or revert back to having specific publish jobs. 

With ccache, we've followed exactly that model. The EFS used for untrusted slaves is a different one than used for the restricted slaves.

Reply:
My concern here is not that people can run random stuff in their Dockerfiles. They can already do that and that's an accepted risk. My point here is that this approach opens up a path for privilige escalation by leveraging the ECR repository to run malicious code in the restricted jobs.

Reply:
Removed the docker push in normal pipelines, now only should run in restricted nightly job.